File: baekjoon/python/1149/1149.py
import sys
sys.stdin = open('1149_input.txt', 'r')

n = int(input())

dp = [[0 for _ in range(3)] for _ in range(n)]
t = []
for _ in range(n):
    t.append(list(map(int, input().split())))

'''
재귀함수로 구현 어떻게 할지 구상
'''
for i in range(n):
    for j in range(3):
        if i == 0:
            dp[i][j] = t[i][j]
        else:
            # [26,40,83]
            # if j=0 : min(40,83) -> min(a[1],a[2])
            # elif j=1 : min(26,83) -> min(a[0],a[2])
            MIN = 1001
            for k in range(3):
                if k == j:
                    continue
                MIN = min(MIN, dp[i-1][k])
            # print(i, j, MIN)

            dp[i][j] = t[i][j] + MIN
            # print(dp)
            # print(t)
            # print()
print(min(dp[n-1]))

File: baekjoon/cpp/1149/1149.cpp
#include <bits/stdc++.h>
using namespace std;

int n;

vector<int> costs[1001];
vector<int> dp[1001];

int main()
{
    std::ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    freopen("1149_input.txt", "r", stdin);

    cin >> n;

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            int temp;
            cin >> temp;
            costs[i].push_back(temp);
            dp[i].push_back(0);
        }
    }

    for (int i = 0; i < 3; i++)
    {
        dp[0][i] = costs[0][i];
    }

    for (int i = 1; i < n; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            int m = 100000000;
            for (int k = 0; k < 3; k++)
            {
                if (j == k)
                    continue;
                else
                {
                    m = dp[i - 1][k] < m ? dp[i - 1][k] : m;
                }
            }
            dp[i][j] = costs[i][j] + m;
        }
    }

    cout << *min_element(dp[n - 1].begin(), dp[n - 1].end()) << "\n";
    return 0;
}
